[AutoDiff] Diagnose duplicate `@differentiable` attributes on properties.

dan-zheng opened the pull request:
- Diagnose duplicate `@differentiable` attributes on computed property
  and getter. Resolves [TF-544](https://bugs.swift.org/browse/TF-544).
- During SILGen, assert no duplicate `@differentiable` attributes.
- Improve duplicate `@differentiable` attribute diagnostics.

dan-zheng committed:
[AutoDiff] Diagnose duplicate `@differentiable` attributes on properties.

- Diagnose duplicate `@differentiable` attributes on computed property
  and getter. Resolves TF-544.
- During SILGen, assert no duplicate `@differentiable` attributes.
- Improve duplicate `@differentiable` attribute diagnostics.

dan-zheng commented on lib/SIL/SILFunctionBuilder.cpp:
```diff
@@ -88,15 +88,23 @@ void SILFunctionBuilder::addFunctionAttributes(SILFunction *F,
           M, indices, A->getRequirements(), M.allocateCopy(jvpName),
           M.allocateCopy(vjpName));
+#if NDEBUG
+      // Verify that no existing attributes have the same indices.
```
Note: addressing [TF-482](https://bugs.swift.org/browse/TF-482) involves checking uniqueness of `@differentiable`/`[differentiable]` attributes based on generic requirements, in addition to parameter indices.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3167,14 +3168,38 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
   }
+  // Transfer `@differentiable` attribute from storage declaration to
+  // getter accessor.
+  if (auto *asd = dyn_cast<AbstractStorageDecl>(D)) {
```
Is it possible to use the original attribute instead? Just set `implicit` to `true`.

dan-zheng replied:
Is that desirable? The original attribute isn't implicit though - I think setting `implicit = true` may affect diagnostics/dumping.

rxwei replied:
In the new attribute you created in line 3185, it's also `implicit`. Did that make a difference to diagnostics or no?

rxwei replied:
I'd suggest writing a computed property with a non-differentiable body, mark the property decl as `@differentiable`, and see what error the AD pass emits.

dan-zheng replied:
Aha - since the original attribute is removed the storage declaration, there's no reason to not mutate/reuse it.

Done in https://github.com/apple/swift/pull/25293/commits/825066220d45c05a58afd3556b78d77f27f234ce.

rxwei commented on test/AutoDiff/differentiable_attr_type_checking.swift:
```diff
@@ -29,16 +29,35 @@ func no_jvp_or_vjp(_ x: Float) -> Float {
+    set { value = newValue }
+  }
+
+  @differentiable(where T == Float) // expected-error {{duplicate '@differentiable' attribute}}
```
This should in theory be valid because generic requirements are different. I remember there are more issues in later phases (e.g. generic requirements not mangled in symbols, etc). If that's why this diagnostic is here, could you please add a FIXME?

dan-zheng replied:
[TF-482](https://bugs.swift.org/browse/TF-482) tracks uniquing `@differentiable` attributes based on generic requirements.
Added TODO comments in https://github.com/apple/swift/pull/25293/commits/6488369c033c263abecb5fa0da22ee7c52c413ff.

dan-zheng committed:
Add TODO comments regarding TF-482.

dan-zheng committed:
Reuse existing `@differentiable` attribute.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3469,7 +3491,9 @@ void AttributeChecker::visitDifferentiatingAttr(DifferentiatingAttr *attr) {
-      diagnoseAndRemoveAttr(da, diag::differentiable_attr_duplicate);
+      TC.diagnose(da->getLocation(), diag::differentiable_attr_duplicate);
+      TC.diagnose(insertion.first->getSecond()->getLocation(),
+                  diag::differentiable_attr_duplicate_note);
```
Do you not need to remove it?

dan-zheng replied:
I thought removing was incorrect for some reason, but removing seems correct. Reverted.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3167,14 +3168,35 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+    if (!insertion.second) {
+      TC.diagnose(attr->getLocation(), diag::differentiable_attr_duplicate);
+      TC.diagnose(insertion.first->getSecond()->getLocation(),
+                  diag::differentiable_attr_duplicate_note);
```
Not remove it?

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3167,14 +3168,35 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
-    diagnoseAndRemoveAttr(attr, diag::differentiable_attr_duplicate);
+    TC.diagnose(attr->getLocation(), diag::differentiable_attr_duplicate);
+    TC.diagnose(insertion.first->getSecond()->getLocation(),
+                diag::differentiable_attr_duplicate_note);
```
Same question here.

dan-zheng committed:
Remove duplicate `@differentiable` attributes.

Revert logic.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Merge remote-tracking branch 'apple/tensorflow' into diagnose-duplicate-diff-attr

dan-zheng commented:
Not sure why two tests failed.
https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/2429/console

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Final fixes.

- Add new `@differentiable` attribute to getter accessor.
  - Using the old attribute caused problems for some reason.
- Make ASTContext `@differentiable` attribute uniquing logic more robust.
- Remove unnecessary SWIFT_ENABLE_TENSORFLOW comment.
- Fix tests.

dan-zheng commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3167,12 +3168,40 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
+    D->getAttrs().removeAttribute(attr);
+    // Transfer `@differentiable` attribute from storage declaration to
+    // getter accessor.
+    auto *newAttr = DifferentiableAttr::create(
```
Note: creating a new attribute here for the getter seems necessary.
I tried reusing `attr` (removing it from the storage declaration `asd` and adding it to `asd->getGetter()`), but that led to problems.

dan-zheng commented on lib/SIL/SILFunctionBuilder.cpp:
```diff
@@ -201,7 +209,6 @@ SILFunctionBuilder::getOrCreateFunction(SILLocation loc, SILDeclRef constant,
       // SWIFT_ENABLE_TENSORFLOW
       addFunctionAttributes(F, storage->getAttrs(), mod, constant);
     }
-    // SWIFT_ENABLE_TENSORFLOW
```
Note: this comment should be removed because the following line is in parity with master.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SIL/SILFunctionBuilder.cpp:
```diff
@@ -88,15 +88,23 @@ void SILFunctionBuilder::addFunctionAttributes(SILFunction *F,
       auto silDiffAttr = SILDifferentiableAttr::create(
           M, indices, A->getRequirements(), M.allocateCopy(jvpName),
           M.allocateCopy(vjpName));
+#if NDEBUG
```
```suggestion
#ifndef NDEBUG
```

dan-zheng replied:
Good catch, thanks!

dan-zheng replied:
Oops,`test/AutoDiff/differentiable_attr_silgen.swift` fails after this change. Investigating.

dan-zheng committed:
Minor fix.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix derived conformances `@differentiable` attribute creation.

Create `@differentiable` attribute on members only if they do not
already have one.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -736,6 +736,11 @@ getOrSynthesizeSingleAssociatedStruct(DerivedConformance &derived,
+      // If member or its getter already has a `@differentiable` attribute,
+      // continue.
+      if (member->getAttrs().hasAttribute<DifferentiableAttr>() ||
+          member->getGetter()->getAttrs().hasAttribute<DifferentiableAttr>())
```
I don't think you need to check the getter decl because users can never write a getter for a stored property.

dan-zheng replied:
Actually, `TypeCheckAttr::visitDifferentiableAttr` transfers `@differentiable` attributes from `AbstractStorageDecl` declarations (including stored properties) to their getters.

The `member->getGetter()->getAttrs().hasAttribute<DifferentiableAttr>()` check is necessary; without it, there's a duplicate attribute assertion violation in SILFunctionBuilder.cpp. I also kept the `member->getAttrs().hasAttribute<DifferentiableAttr>()` check for robustness.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
